### PR TITLE
1328 - Create Booking from Placement Request

### PR DESCRIPTION
**Pull request by brownbearsoftware**

### CODE REVIEW

**Comment by brownbearsoftware** [comment]
[Add an endpoint to create a booking from a placement request](https://trello.com/c/SjH2GSPI/1328-add-an-endpoint-to-create-a-booking-from-a-placement-request)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/results/AuthorisableActionResult.kt [hunk 1]
```diff
@@ -3,5 +3,5 @@ package uk.gov.justice.digital.hmpps.approvedpremisesapi.results
 sealed interface AuthorisableActionResult<EntityType> {
   data class Success<EntityType>(val entity: EntityType) : AuthorisableActionResult<EntityType>
   class Unauthorised<EntityType> : AuthorisableActionResult<EntityType>
-  class NotFound<EntityType> : AuthorisableActionResult<EntityType>
+  class NotFound<EntityType>(val entityType: String? = null, val id: String? = null) : AuthorisableActionResult<EntityType>
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/results/AuthorisableActionResult.kt:6]: 🙌 I wanted to do this for something else, but I lacked the technical nous!

File: src/main/resources/static/api.yml [hunk 2]
```diff
@@ -2099,6 +2099,39 @@ paths:
           $ref: '#/components/responses/403Response'
         500:
           $ref: '#/components/responses/500Response'
+  /placement-requests/{id}/booking:
```
Comment (pezholio) [inline src/main/resources/static/api.yml:2102]: Thinking about this - I wonder if it makes more sense to relate the booking to the Application, as this is the thing we _really_ care about. The PlacementRequest is more a record to aid with matching

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt [hunk 3]
```diff
@@ -82,12 +82,11 @@ class BookingService(
   private val cancellationReasonRepository: CancellationReasonRepository,
   @Value("\${application-url-template}") private val applicationUrlTemplate: String
 ) {
-  fun createBooking(bookingEntity: BookingEntity): BookingEntity = bookingRepository.save(bookingEntity)
   fun updateBooking(bookingEntity: BookingEntity): BookingEntity = bookingRepository.save(bookingEntity)
   fun getBooking(id: UUID) = bookingRepository.findByIdOrNull(id)
 
   @Transactional
-  fun createApprovedPremisesBooking(
+  fun createApprovedPremisesAdHocBooking(
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt:166]: Good naming - we might want to bin this off at some point I guess

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt [hunk 4]
```diff
@@ -82,11 +83,85 @@ class BookingService(
...
   private val cancellationReasonRepository: CancellationReasonRepository,
   private val bedRepository: BedRepository,
+  private val placementRequestRepository: PlacementRequestRepository,
   @Value("\${application-url-template}") private val applicationUrlTemplate: String
 ) {
   fun updateBooking(bookingEntity: BookingEntity): BookingEntity = bookingRepository.save(bookingEntity)
   fun getBooking(id: UUID) = bookingRepository.findByIdOrNull(id)
 
+  @Transactional
+  fun createApprovedPremisesBookingFromPlacementRequest(
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt:93]: Again, sorry for the misdirection here, but I reckon the `createApprovedPremisesBookingFromApplication` is the thing to do here

**Review by pezholio — COMMENTED** [review]
All in all, this looks great, but (sorry!) I think an Application might be the best first-class relation to use

File: src/main/resources/static/api.yml (hunk 2 above)
Comment (pezholio) [reply src/main/resources/static/api.yml:2102]: On reflection (and chatting to @brownbearsoftware) I realised that Applications can have multiple PlacementRequests, so linking through the request is the thing to do here

**Pull request merged**